### PR TITLE
1870: river division, special tiles upgrade path, share redemption and reissue

### PR DESCRIPTION
**Pull request by daniel-sousa-me**

### CODE REVIEW

File: assets/app/view/game/part/divisions.rb [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            a_control = VERTICES[(division.a + division.asign) % 6]
+            vertex_a = convex_combination(VERTICES[division.a], a_control)
+            b_control = VERTICES[(division.b + division.bsign) % 6]
+            vertex_b = convex_combination(VERTICES[division.b], b_control)
+
+            da = if division.asign.nonzero?
+                   VERTICES[division.a].map { |x| x * (1 - (1 - COEFFICIENT) * 2) } # cos(30) = 1/2
+                 else
+                   convex_combination(vertex_a, vertex_b)
+                 end. join(' ')
```
Comment (tobymao) [inline assets/app/view/game/part/divisions.rb:66]: shouldn't be a space

File: lib/engine/game/base.rb [hunk 2]
```diff
@@ -1027,7 +1027,7 @@ def upgrades_to?(from, to, special = false)
         return true if special
 
         # correct label?
-        return false if from.label != to.label
+        return false unless from.label == to.label || (!from.label && to.label&.optional_label?)
```
Comment (tobymao) [inline lib/engine/game/base.rb:1030]: don't use unless with multiple conditions

File: lib/engine/ability/blocks_division.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+require_relative 'base'
+
+module Engine
+  module Ability
+    class BlocksDivision < Base
+      attr_reader :division_type
```
Comment (tobymao) [inline lib/engine/ability/blocks_division.rb:8]: just call it type

File: lib/engine/part/division.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+require_relative 'base'
+
+module Engine
+  module Part
+    class Division < Base
+      attr_reader :a, :asign, :b, :bsign, :type, :restrict, :magnet, :blockers, :inner, :outer
```
Comment (tobymao) [inline lib/engine/part/division.rb:8]: explain everything here, also it should probably be a_sign and b_sign

File: lib/engine/ability/blocks_division.rb (hunk 3 above)
Comment (daniel-sousa-me) [reply lib/engine/ability/blocks_division.rb:8]: `type` is already used for the type of the ability

**Comment by daniel-sousa-me** [comment]
I keep confusing what's from this PR and what's I'm working on. I'm a bit tired of rebasing 5 times every day. I'm marking this as draft for now because it got messed up. Since I'm almost finished implementing everything anyway, I'll resubmit once it's done.

**Comment by tobymao** [comment]
closing for now

**Pull request closed without merging**